src/shims/io.rs

```diff
@@ -0,0 +1,213 @@
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+pub struct FileHandle {
+    file: File,
+    flag: i32,
+}
+
+pub struct FileHandler {
+    handles: HashMap<i32, FileHandle>,
+    low: i32,
+}
+
+impl Default for FileHandler {
+    fn default() -> Self {
+        FileHandler {
+            handles: Default::default(),
+            // 0, 1 and 2 are reserved for stdin, stdout and stderr
+            low: 3,
+        }
+    }
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn open(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        flag_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
+
+        if flag != this.eval_libc_i32("O_RDONLY")? && flag != this.eval_libc_i32("O_CLOEXEC")? {
+            throw_unsup_format!("Unsupported flag {:#x}", flag);
+        }
+
+        let path_bytes = this
+            .memory()
+            .read_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+        let path = std::str::from_utf8(path_bytes)
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?;
+        let fd = File::open(path).map(|file| {
+            let mut fh = &mut this.machine.file_handler;
+            fh.low += 1;
+            fh.handles.insert(fh.low, FileHandle { file, flag });
+            fh.low
+        });
+
+        this.consume_result(fd)
+    }
+
+    fn fcntl(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        cmd_op: OpTy<'tcx, Tag>,
+        arg_op: Option<OpTy<'tcx, Tag>>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd == this.eval_libc_i32("F_SETFD")? {
+            // This does not affect the file itself. Certain flags might require changing the file
+            // or the way it is accessed somehow.
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            // The only usage of this in stdlib at the moment is to enable the `FD_CLOEXEC` flag.
+            let fd_cloexec = this.eval_libc_i32("FD_CLOEXEC")?;
+            if let Some(FileHandle { flag: old_flag, .. }) =
+                this.machine.file_handler.handles.get_mut(&fd)
+            {
+                if flag ^ *old_flag == fd_cloexec {
+                    *old_flag = flag;
+                } else {
+                    throw_unsup_format!("Unsupported arg {:#x} for `F_SETFD`", flag);
```

Comment:
Note that unlike in the `open` case, it is far from obvious that this errors if e.g. any bit other than `fd_cloexec` is set... in case that's what you meant?

Reply:
Yes I understand what you mean, I'm not completely happy either with this part of the code either. 

The thing is that `std::fs` sets this flag for some reason but AFAIK that flag only matters if you're executing the file and I couldn't find anything inside the standard library where using this flag actually matters.

So yeah my approach was to only allow `F_SETFD` if it was used for setting the `FD_CLOEXEC` flag. Maybe I should change the error to `"The only supported argument for `F_SETFD` is `FD_CLOEXEC`" ` or something like that.

Reply:
That would work if the file is in read only mode (because someone decided `0` was a great flag value). But for example if its in append or write mode, `flags == fd_cloexec` would return false

Reply:
Ah so it does have all the old flags still in there, and just adds a new one. I see.

Reply:
yes so the point is that the difference between the flags must be exactly the cloexec one.

Reply:
But even then, miri is supposed to let `fcntl(fd, F_SETF, FD_CLOEXEC)` run because that's used by `std::fs` somewhere (sorry I don't remember where actually)

Reply:
pfff I'm an idiot... the easy way to do this is to check that `fcntl(fd, F_SETF, x)` is called only if `x == FD_CLOEXEC`.

Reply:
> But even then, miri is supposed to let fcntl(fd, F_SETF, FD_CLOEXEC) run because that's used by std::fs somewhere (sorry I don't remember where actually)

It is used only if `F_GETFD` returns that `FD_CLOEXEC` is not set.

If you make `F_GETFD` always just return `FD_CLOEXEC`, there should be no more calls to `FD_SETFD` at all.

Reply:
It is OK to just return the actual flags for the file descriptor because the standard library always opens files with `FD_CLOEXEC` enabled: https://github.com/rust-lang/miri/pull/987

Reply:
Which flags? The flags you are storing in the table are the `open` flags, not to be confused with the `fcntl` flags! libstd always sets `O_CLOEXEC` with `open`, and that *should* have the effect of `FD_CLOEXEC` being set when doing `F_GETFD`. That's why I went on about those two families of flags and even [wrote a testcase](https://github.com/rust-lang/miri/pull/962#discussion_r333901667) for you. ;)

Do not confuse the internal flags of the file with the ways these flags are controlled via syscalls. Right now, there is only one flag (on Linux) that a file internally has: `FD_CLOEXEC`. Its initial state is determined by whether `O_CLOEXEC` is set. You *could* have a `bool` for that in the table if you want, that would let us support some programs that use libc APIs directly (and not just through libstd).

All the other `O_*` flags are forgotten after the file has been opened, they are not needed any more.

Reply:
yes I just made my mind about it. I changed this here: https://github.com/rust-lang/miri/pull/987